morango/api/viewsets.py

```diff
@@ -338,83 +383,88 @@ def create(self, request):  # noqa: C901
         if scope_error_msg:
             return response.Response(scope_error_msg, status=status.HTTP_403_FORBIDDEN)
 
-        # build the data to be used for creating the transfersession
-        data = {
-            "id": request.data.get("id"),
-            "start_timestamp": timezone.now(),
-            "last_activity_timestamp": timezone.now(),
-            "active": True,
-            "filter": requested_filter,
-            "push": is_a_push,
-            "records_total": request.data.get("records_total") if is_a_push else None,
-            "sync_session": syncsession,
-            "client_fsic": request.data.get("client_fsic") or "{}",
-            "server_fsic": "{}",
-        }
-
-        transfersession = core.TransferSession(**data)
-        transfersession.full_clean()
-        transfersession.save()
-
-        # must update database max counters before calculating fsics
-        if not is_a_push:
-
-            if getattr(settings, "MORANGO_SERIALIZE_BEFORE_QUEUING", True):
-                with OperationLogger("Serializing records", "Serialization complete"):
-                    _serialize_into_store(
-                        transfersession.sync_session.profile, filter=requested_filter
-                    )
+        context = LocalSessionContext(
+            request=request,
+            sync_session=syncsession,
+            sync_filter=requested_filter,
+            is_push=is_a_push,
+        )
 
-        transfersession.server_fsic = json.dumps(
-            DatabaseMaxCounter.calculate_filter_max_counters(requested_filter)
+        # If both client and ourselves allow async, we just return accepted status, and the client
+        # should PATCH the transfer_session to the appropriate stage. If not async, we wait until
+        # queuing is complete
+        to_stage = (
+            transfer_stage.INITIALIZING
+            if self.async_allowed()
+            else transfer_stage.QUEUING
         )
-        transfersession.save()
+        result = session_controller.proceed_to_and_wait_for(to_stage, context=context)
 
-        if not is_a_push:
+        if result == transfer_status.ERRORED:
+            return response.Response(
+                "Failed to initialize session",
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            )
```

Comment:
I recently added `last_error` to the controller, but overall I think conveying errors could be improved. Quick win is to use `last_error` here, but I think ideally, we should move that to the context since having it on the controller won't work for truly async operations.